File: sorento_crm_frontend/app/(protected)/scm/reorder/lib/coverageTimeline.ts
/**
 * Coverage Timeline - PURE display helpers (UAC Group B). No React, no fetching.
 *
 * The one rule this file exists to honour: **months are display grouping only**
 * (AC-B3). Grouping never touches a balance - `groupRowsByMonth` re-partitions the
 * rows the server already accumulated and returns them untouched, so no arithmetic
 * can hide in the presentation layer.
 *
 * Survives Phase 2. Only `coverageMockStore.ts` is deleted.
 */
import type { CoverageRow, CoverageShortfall } from '../types/coverage.types';

/** One month heading plus the rows underneath it. `key` is null for the opening row. */
export interface CoverageMonthGroup {
  /** `YYYY-MM`, or null for the undated opening row. */
  key: string | null;
  /** Rendered heading, e.g. "Jul 2026". */
  label: string;
  rows: CoverageRow[];
}

const MONTH_FMT = new Intl.DateTimeFormat('en-MY', { month: 'short', year: 'numeric' });

const DAY_FMT = new Intl.DateTimeFormat('en-MY', {
  day: '2-digit',
  month: 'short',
  year: 'numeric',
  timeZone: 'UTC',
});

/** `2026-07-01` -> `Jul 2026`. Parsed as a plain date, so no timezone shift. */
export function monthLabel(isoDate: string): string {
  const [y, m] = isoDate.split('-').map(Number);
  if (!y || !m) return isoDate;
  return MONTH_FMT.format(new Date(Date.UTC(y, m - 1, 1)));
}

/**
 * `2026-08-03` -> `03 Aug 2026`.
 *
 * Built on `Date.UTC` and formatted in UTC on purpose. An event date is a plain
 * calendar date with no time and no zone; running it through a local-timezone
 * formatter moves a 1 Jul required-date to 30 Jun west of Greenwich, which would
 * silently reorder the timeline against the balances the server accumulated.
 */
export function dayLabel(isoDate: string | null): string {
  if (!isoDate) return '';
  const [y, m, d] = isoDate.slice(0, 10).split('-').map(Number);
  if (!y || !m || !d) return isoDate;
  return DAY_FMT.format(new Date(Date.UTC(y, m - 1, d)));
}

/**
 * `2026-08-03T09:12:00` -> `03 Aug 2026, 09:12`.
 *
 * Rendered as written, with NO conversion: `computed_at` is already naive Malaysia
 * wall-clock (see `coverage.types.ts`). Passing it to `formatDateTimeInMalaysia`,
 * which reads a naive string as UTC, would add eight hours to a timestamp that is
 * already local.
 */
export function computedAtLabel(iso: string | null): string {
  if (!iso) return '';
  const [datePart, timePart] = iso.split('T');
  const day = dayLabel(datePart);
  const time = (timePart ?? '').slice(0, 5);
  return time ? `${day}, ${time}` : day;
}

/**
 * Partition rows under month headings, preserving order and every balance.
 * Mirrors the backend's `group_by_month` so the two cannot drift.
 */
export function groupRowsByMonth(rows: CoverageRow[]): CoverageMonthGroup[] {
  const out: CoverageMonthGroup[] = [];
  for (const row of rows) {
    const at = row.event.at;
    const key = at ? at.slice(0, 7) : null;
    const last = out[out.length - 1];
    if (last && last.key === key) {
      last.rows.push(row);
      continue;
    }
    out.push({
      key,
      label: key ? monthLabel(`${key}-01`) : 'Opening',
      rows: [row],
    });
  }
  return out;
}

/** Whole days between two ISO dates, ignoring time of day. */
export function daysBetween(fromIso: string, toIso: string): number | null {
  const from = Date.parse(`${fromIso.slice(0, 10)}T00:00:00Z`);
  const to = Date.parse(`${toIso.slice(0, 10)}T00:00:00Z`);
  if (Number.isNaN(from) || Number.isNaN(to)) return null;
  return Math.round((to - from) / 86_400_000);
}

/** The supply that arrives too late, and by how many days. */
export interface LateSupply {
  ref: string;
  label: string;
  at: string;
  qty: number;
  days_late: number;
}

/**
 * The first supply event dated AFTER the shortfall (AC-B4).
 *
 * This is what makes the gap legible instead of netted away: a PO of 200 arriving
 * 25 Aug against an order due 3 Aug leaves the closing balance positive while the
 * order is still short, and the number a planner needs is the 22 days between them.
 * Reads the rows the server produced; computes no balance.
 */
export function firstLateSupply(
  rows: CoverageRow[],
  shortfall: CoverageShortfall | null,
): LateSupply | null {
  if (!shortfall?.at) return null;
  for (const row of rows) {
    const e = row.event;
    if (e.kind !== 'supply' || !e.at) continue;
    const gap = daysBetween(shortfall.at, e.at);
    if (gap === null || gap <= 0) continue;
    return { ref: e.ref, label: e.label, at: e.at, qty: e.qty, days_late: gap };
  }
  return null;
}

/**
 * The incoming supply on the timeline, split by stage.
 *
 * Their SUM is what the running balance already used; the SPLIT is what a person
 * deciding a quantity reads, because only the on-order half can still be re-dated
 * or cancelled. Derived from the rows the server produced, so it can never disagree
 * with the balance.
 */
export function supplySplit(rows: CoverageRow[]): { on_order: number; in_transit: number } {
  let onOrder = 0;
  let inTransit = 0;
  for (const row of rows) {
    const e = row.event;
    if (e.kind !== 'supply') continue;
    if (e.supply_stage === 'in_transit') inTransit += e.qty;
    else onOrder += e.qty;
  }
  return { on_order: onOrder, in_transit: inTransit };
}

/** Short human name for a cover source. */
export const SOURCE_LABEL: Record<string, string> = {
  own: 'Own bin',
  brw: 'Shared pool',
  other_project: "Another holder's bin",
  order: 'Purchase',
};

/** Short human name for a supply stage. */
export const SUPPLY_STAGE_LABEL: Record<string, string> = {
  on_order: 'On order',
  in_transit: 'In transit',
};

/**
 * The one-line verdict a planner acts on: use existing stock, or buy.
 * `use_stock` true is the SRTWT7408 case - demand 67 against a pool holding 4,397
 * must read "use the pool", never a buy of 67 (AC-B1c).
 *
 * It answers ONE question: is the committed demand covered from stock we already hold.
 * The reorder engine on the same screen answers a different one: is stock above its
 * reorder point. Those disagree routinely, so when the demand is covered AND the balance
 * is under the floor, the verdict states its own scope and reports the other answer
 * beside it. Left unqualified it read "existing stock covers it" next to a grid row
 * recommending a buy of 18,551, and the sentence gets believed over the number.
 */
export function coverVerdict(
  useStock: boolean,
  buyQty: number,
  poolLocation: string,
  shortfall: CoverageShortfall | null = null,
): { tone: 'stock' | 'buy'; headline: string; note: string | null } {
  if (useStock) {
    if (shortfall) {
      const short = Math.round(shortfall.qty).toLocaleString('en-MY');
      return {
        tone: 'stock',
        headline: 'Committed demand is covered',
        note: `Below the reorder point by ${short}${
          poolLocation ? ` across ${poolLocation}` : ''
        }`,
      };
    }
    return {
      tone: 'stock',
      headline: poolLocation ? `Use the pool (${poolLocation})` : 'Use existing stock',
      note: null,
    };
  }
  return { tone: 'buy', headline: `Buy ${buyQty.toLocaleString('en-MY')}`, note: null };
}

File: sorento_crm_backend/app/services/scm/coverage_timeline.py
"""Coverage Timeline: the dated running balance, and where a demand line's stock comes from.

PURE maths, no I/O, golden-testable in isolation - the same discipline as
``reorder_engine``'s top half and ``cash_ranking``.

Two things live here.

**1. The dated running balance.** Every dated event (opening on hand, each committed demand
line at its required date, each open supply line at its expected date) sorted by date and
accumulated. The first event that pushes the balance below the floor is the shortfall, and
that event's date is the need-by date.

There are deliberately NO time buckets in the arithmetic. Months are a display grouping
only (`group_by_month`). Aggregating a month would let supply arriving on the 25th cover
demand due on the 3rd, so the report would understate shortfall - the one class of error
that destroys trust in a planning tool in week one. See ADR-0011.

**Why this does not endanger the M3 golden set.** The blessed goldens cover the pure
formulae (`safety_stock`, `reorder_point`, `order_up_to`, `trigger`, `order_qty`) and feed
them plain values. Those functions are untouched. What changes is where `net` comes from:
`closing_balance` over a timeline is arithmetically identical to
``on_hand + on_order - committed`` when every event shares one date, which is exactly the
continuous-replenishment case. So ROP is the one-bucket degenerate case of this, and the
parity is structural rather than coincidental.

**2. Source resolution.** A shortage is not automatically a purchase. A demand line is
covered from its own reserved bin, then the shared pool at the same site, then another
holder's bin (which needs their agreement), and only then bought. Those four are the
existing ``so_line_allocations.source_type`` values, reused rather than reinvented.

The case that forced this: item SRTWT7408 carries demand of 67 against customer bins while
4,397 sit in the BRW pool. Netting per warehouse reports a shortfall and recommends buying
67 units of an item holding 4,397 - on the first row of the first report anyone reads.
"""
from __future__ import annotations

import itertools
from dataclasses import dataclass, field
from datetime import date
from typing import Iterable, Optional, Sequence

# --------------------------------------------------------------------------- #
# event kinds and source types
# --------------------------------------------------------------------------- #

# The one threshold that governs BOTH the shortfall trigger and the reported figure.
# They used to differ: the trigger was 1e-9 while the quantity was rounded to 4 decimals,
# so any gap in (1e-9, 5e-5) reported "Short 0" beside a peak deficit of 0 and the hint
# "never goes short". Since the floor is a computed reorder point (12369.655534 in the
# fixtures), that window is reachable. Rounding first, then comparing on the SAME
# precision, is what makes the two figures agree by construction.
QTY_PRECISION = 4
EPSILON = 10 ** -QTY_PRECISION / 2

OPENING = "opening"
DEMAND = "demand"
SUPPLY = "supply"

# Mirrors so_line_allocations.source_type in the project-sales work. Do NOT invent a
# parallel vocabulary: one of these strings ends up persisted on the allocation.
SOURCE_OWN = "own"
SOURCE_POOL = "brw"
SOURCE_OTHER = "other_project"
SOURCE_ORDER = "order"

# Incoming stock is the SPO allocation against a shipment, and it is the ONLY supply the
# balance counts (decision, 6 Aug 2026). `SUPPLY_ON_ORDER` is no longer emitted by any
# producer: a purchase order is an order placed, not stock on the water, so its quantity is
# reported beside the timeline (`Coverage.qty_ordered_not_incoming`) and never on it. The
# constant stays because persisted rows and API consumers still carry the string.
SUPPLY_ON_ORDER = "on_order"
SUPPLY_IN_TRANSIT = "in_transit"


@dataclass(frozen=True)
class TimelineEvent:
    """One dated movement. ``qty`` is signed: negative consumes, positive replenishes."""

    at: Optional[date]          # None only for the opening balance
    qty: float
    kind: str                   # opening | demand | supply
    ref: str = ""               # document number, shown on screen. Never a UUID.
    label: str = ""             # human context: project or customer or supplier
    location: str = ""          # the bin or pool the movement sits in
    supply_stage: Optional[str] = None   # in_transit, for SUPPLY only (see above)
    # The last two keys of PLAN-scm-front-planning.md 3.5's ordering. `line_no` is the
    # human line number on the document (missing sorts last); `line_id` is the internal
    # SO-line id, a final stable key that is NEVER displayed - two rows agreeing on
    # everything a person can see still have to sort the same way on every run.
    line_no: Optional[int] = None
    line_id: Optional[str] = None


@dataclass(frozen=True)
class TimelineRow:
    event: TimelineEvent
    balance: float


@dataclass(frozen=True)
class Shortfall:
    """The first point the balance falls below the floor.

    Carries a date as well as a quantity because a quantity alone cannot be acted on:
    buying it late is indistinguishable from not buying it.
    """

    at: Optional[date]
    qty: float                  # how much is missing at that point, positive
    ref: str                    # the document that tipped it
    label: str = ""


@dataclass(frozen=True)
class TimelineResult:
    rows: list[TimelineRow]
    closing_balance: float
    shortfall: Optional[Shortfall]
    # The worst the balance ever gets, which is what has to be covered. Distinct from the
    # first shortfall's own gap: a later event can dig deeper, and buying only the first
    # gap would leave the plan short a second time.
    peak_deficit: float
    # How many events fell beyond the horizon and are therefore absent from `rows`. An
    # omission a screen does not mention reads as data that is not there, so the count
    # travels with the result rather than the caller having to re-derive it.
    excluded_event_count: int = 0


def _sort_key(e: TimelineEvent) -> tuple:
    """PLAN-scm-front-planning.md 3.5's ordering, and the ONLY one any consumer uses.

    Date, then kind with SUPPLY BEFORE DEMAND, then SO number, then line number with
    missing numbers last, then the internal line id as a final stable key.

    Supply-first on a same-day tie is Stage 1C's deliberate reversal of the earlier
    demand-first order. Incoming SPO is dated location supply, and an SPO arriving on a
    line's required date "counts at that date" (Q2): a container that clears and is put
    away in the morning covers a despatch due the same day, and treating the despatch as
    though the stock never arrived reports a shortfall that does not exist. There is one
    ordering contract - the reorder engine, the Coverage Timeline panel, plan exceptions
    and the coverage routes all inherit this, with no per-consumer divergence.
    """
    return (
        e.at or date.min,
        0 if e.kind == OPENING else (1 if e.kind == SUPPLY else 2),
        e.ref,
        e.line_no is None,
        e.line_no if e.line_no is not None else 0,
        e.line_id or "",
    )


def build_timeline(
    opening: float,
    events: Iterable[TimelineEvent],
    *,
    floor: float = 0.0,
    horizon_end: Optional[date] = None,
) -> TimelineResult:
    """Accumulate dated events into a running balance and find the first shortfall.

    ``floor`` is the level the balance must not fall below - 0 for project demand, or the
    reorder point for a continuous SKU, which is what makes ROP the one-bucket case.

    ``horizon_end`` drops events beyond the planning horizon. Excluded events are absent
    from ``rows`` but COUNTED into ``excluded_event_count``: a bound is necessary (a
    ten-year tail makes the report unreadable) and the count is what keeps it honest,
    because a planner who cannot see that nine years of demand was dropped reads the
    visible shortfall as the whole picture.
    """
    kept: list[TimelineEvent] = []
    excluded = 0
    for e in events:
        if horizon_end is not None and e.at is not None and e.at > horizon_end:
            excluded += 1
            continue
        kept.append(e)
    ordered = sorted(kept, key=_sort_key)

    rows: list[TimelineRow] = [
        TimelineRow(
            event=TimelineEvent(at=None, qty=0.0, kind=OPENING, ref="", label="opening on hand"),
            balance=float(opening),
        )
    ]
    balance = float(opening)
    shortfall: Optional[Shortfall] = None
    peak_deficit = max(0.0, floor - balance)
    # Stock ALREADY under the floor is a shortfall, and the most urgent kind: it is short
    # now rather than on some future date. Seeding only `peak_deficit` from it left the two
    # figures disagreeing about whether anything was wrong, and a screen that prints
    # "nothing is short" beside a non-zero deficit gets believed over the number.
    # `at` stays None because no event caused it and the opening balance carries no date;
    # stamping today would assert a fact the data does not hold.
    peak_deficit = round(peak_deficit, QTY_PRECISION)
    if peak_deficit >= EPSILON:
        shortfall = Shortfall(
            at=None, qty=peak_deficit, ref="", label="opening on hand"
        )

    for e in ordered:
        balance += float(e.qty)
        rows.append(TimelineRow(event=e, balance=balance))
        gap = round(floor - balance, QTY_PRECISION)
        if gap >= EPSILON:
            if shortfall is None:
                shortfall = Shortfall(at=e.at, qty=gap, ref=e.ref, label=e.label)
            peak_deficit = max(peak_deficit, gap)

    return TimelineResult(
        rows=rows,
        closing_balance=round(balance, 4),
        shortfall=shortfall,
        peak_deficit=peak_deficit,
        excluded_event_count=excluded,
    )


def closing_balance(opening: float, events: Iterable[TimelineEvent]) -> float:
    """The dateless net position, expressed through the timeline.

    This is the bridge that makes ROP the one-bucket case: with every event on one date
    this returns exactly ``on_hand + on_order - committed``, which is what
    ``scm.net_position_v`` computes and what `trigger` and `order_qty` already consume.
    Those formulae are therefore untouched by the date axis.
    """
    return build_timeline(opening, events).closing_balance


def group_by_month(rows: Sequence[TimelineRow]) -> list[tuple[Optional[str], list[TimelineRow]]]:
    """Group rows under month headings for display. Presentation only.

    Changing the grouping cannot change a balance, which is the property that makes the
    no-buckets decision safe to render as if it had buckets.
    """
    out: list[tuple[Optional[str], list[TimelineRow]]] = []
    for key, grp in itertools.groupby(
        rows, key=lambda r: r.event.at.strftime("%Y-%m") if r.event.at else None
    ):
        out.append((key, list(grp)))
    return out


# --------------------------------------------------------------------------- #
# source resolution: a shortage is not automatically a purchase
# --------------------------------------------------------------------------- #


@dataclass(frozen=True)
class SourceAvailability:
    """What could cover a demand line, by source, at the moment of asking.

    Computed live and never stored: a snapshot of somebody else's on-hand is stale the
    moment they ship, and acting on a stale figure is the failure this exists to prevent.
    """

    own: float = 0.0                     # the line's own reserved bin
    pool: float = 0.0                    # shared pool at the SAME site
    other: float = 0.0                   # other holders' bins, needs their agreement
    pool_location: str = ""
    other_locations: tuple[str, ...] = field(default_factory=tuple)


@dataclass(frozen=True)
class SourceAllocation:
    source_type: str
    qty: float
    location: str = ""
    needs_claim: bool = False


def resolve_sources(
    demand_qty: float,
    availability: SourceAvailability,
    *,
    allow_borrow: bool = True,
) -> list[SourceAllocation]:
    """Cover a demand quantity from the cheapest source outward.

    Order is own bin, then same-site pool, then another holder's bin behind a claim, then
    buy. Only the last needs a quantity and a supplier, which is the whole point: treating
    "buy" as the only answer to a shortage is what would recommend purchasing an item that
    is already sitting in the pool.

    ``allow_borrow=False`` skips the claim step for callers that will not chase another
    holder - the remainder becomes a purchase instead, which is a slower but honest answer.

    A cross-site pool is deliberately NOT a source. Stock at another site is real but needs
    a physical movement with a cost and a lead time, so covering from it is a separate
    proposal a person accepts, never a silent netting.
    """
    remaining = max(0.0, float(demand_qty))
    out: list[SourceAllocation] = []

    for qty_available, source_type, location, needs_claim in (
        (availability.own, SOURCE_OWN, "", False),
        (availability.pool, SOURCE_POOL, availability.pool_location, False),
        (availability.other if allow_borrow else 0.0, SOURCE_OTHER,
         ", ".join(availability.other_locations), True),
    ):
        if remaining <= 1e-9:
            break
        take = min(remaining, max(0.0, float(qty_available)))
        if take <= 1e-9:
            continue
        out.append(
            SourceAllocation(
                source_type=source_type,
                qty=round(take, 4),
                location=location,
                needs_claim=needs_claim,
            )
        )
        remaining -= take

    if remaining > 1e-9:
        out.append(SourceAllocation(source_type=SOURCE_ORDER, qty=round(remaining, 4)))

    return out


def buy_quantity(allocations: Sequence[SourceAllocation]) -> float:
    """How much of a resolved demand actually has to be bought. Zero means use stock."""
    return round(sum(a.qty for a in allocations if a.source_type == SOURCE_ORDER), 4)


def is_use_stock(allocations: Sequence[SourceAllocation]) -> bool:
    """True when existing stock covers the whole line, i.e. Mr Loo's "BRW" answer."""
    return buy_quantity(allocations) <= 1e-9
